### PR TITLE
change test config storage path from extension folder to storage folder

### LINKED ISSUE #154: Config file can't auto-generate project configuration information
1. Open A project
2. Open config file by triggering command  `Java: Edit test configuration`
3. The config file will generate configuration content for A project
4. Close A project and open B project
5. Open config file again by triggering command  `Java: Edit test configuration` 

Result: 
Configuration content without any updates

### PR DESCRIPTION
**Pull request by ansyral**
Signed-off-by: xuzho <xuzho@microsoft.com>

to fix #154 #153

### CODE REVIEW

File: extension/src/testConfigManager.ts [hunk 1]
```diff
@@ -11,7 +12,9 @@ import { TestConfig } from './Models/testConfig';
 import * as Logger from './Utils/Logger/logger';
 
 export class TestConfigManager {
-    constructor(private readonly _configPath: string, private readonly _projectManager: ProjectManager) {
+    private readonly _configPath: string;
+    constructor(storagePath: string, private readonly _projectManager: ProjectManager) {
+        this._configPath = path.join(storagePath, 'configs', Configs.TEST_LAUNCH_CONFIG_NAME);
```
Comment (yaohaizh) [inline extension/src/testConfigManager.ts]: If make this config file locate at the project specific user folder, how can user make changes to this file?

Comment (ansyral) [reply extension/src/testConfigManager.ts]: we provide command `Java: Edit Test Configuration` for user to open config, but this is a good suggestion, I would move the config to `.vscode` folder where user could directly access it.

File: extension/src/extension.ts [hunk 2]
```diff
@@ -193,7 +193,13 @@ function runTest(tests: TestSuite[] | TestSuite, isDebugMode: boolean, config?:
 }
 
 async function getTestConfig(configManager: TestConfigManager, isDebugMode: boolean): Promise<RunConfig> {
-    const config: TestConfig = await configManager.loadConfig();
+    let config: TestConfig;
+    try {
+        config = await configManager.loadConfig();
+    } catch (ex) {
+        window.showErrorMessage('Failed to load test config! Please check whether your test configuration is a valid JSON file');
```
Comment (yaohaizh) [inline extension/src/extension.ts]: `the` test config. Same below

File: extension/src/testConfigManager.ts [hunk 3]
```diff
@@ -11,46 +12,45 @@ import { TestConfig } from './Models/testConfig';
...
 
 export class TestConfigManager {
-    constructor(private readonly _configPath: string, private readonly _projectManager: ProjectManager) {
+    private readonly _configPath: string;
+    constructor(private readonly _projectManager: ProjectManager) {
+        const workspaceFolders = workspace.workspaceFolders;
+        if (!workspaceFolders) {
+            throw new Error('Not supported without a folder!');
+        }
+        this._configPath = path.join(workspaceFolders[0].uri.fsPath, '.vscode', Configs.TEST_LAUNCH_CONFIG_NAME);
```
Comment (yaohaizh) [inline extension/src/testConfigManager.ts]: How about multiple root scenario?  Always pick up the first one?

File: extension/src/testConfigManager.ts [hunk 4]
```diff
@@ -59,6 +59,34 @@ export class TestConfigManager {
         });
     }
 
+    private createTestConfigIfNotExisted(): Promise<void> {
+        return new Promise((resolve, reject) => {
+            mkdirp(path.dirname(this._configPath), (merr) => {
+                if (merr && merr.code !== 'EEXIST') {
```
Comment (yaohaizh) [inline extension/src/testConfigManager.ts]: Storage path should be `File path` ?

File: extension/src/testConfigManager.ts (hunk 3 above)
Comment (ansyral) [reply extension/src/testConfigManager.ts]: yes , I would always store it in the first one considering the config for different projects are merged in one config

**Pull request merged**